lib/ably/realtime/channel/channel_manager.rb

```diff
@@ -199,14 +202,18 @@ def channel_retry_timeout
       end
 
       def send_attach_protocol_message
-        send_state_change_protocol_message Ably::Models::ProtocolMessage::ACTION.Attach, :suspended # move to suspended
+        message_opptions = {}
+        message_opptions[:flags] = channel.options.modes_to_flags if channel.options.modes
+        message_opptions[:params] = channel.options.params if channel.options.params.any?
```

Comment:
@lukaszsliwa well. If you mean a situation where we
```
1. Set params
2. Send message
3. Unset params
```
It will work as expected as this message_options are set only for one message sending

Reply:
It does not work. For example:

```ruby
channel_options.set_params({ occupancy: true })
channel_options.params
# { occupancy: true }

channel_options.set_params({})
channel_options.params
# { occupancy: true } # WRONG. It should return {} (empty hash)

```

Unit test for `ChannelOptions#set_params` is missing.

Here is the case and example unit test:

```ruby
# spec/lib/unit/models/channel_options_spec.rb

# frozen_string_literal: true

require 'spec_helper'

RSpec.describe Ably::Models::ChannelOptions do
  let(:modes) { nil }
  let(:params) { {} }
  let(:options) { described_class.new(modes: modes, params: params) }
  # ...

  describe '#set_params' do
    let(:previous_params) { { example_attribute: 1 } }
    let(:new_params) { { new_attribute: 1 } }
    let(:params) { previous_params }

    it 'should be able to overwrite attributes' do
      expect { options.set_params(new_params) }.to \
        change { options.params }.from(previous_params).to(new_params)
    end

    it 'should be able to make params empty' do # (1)
      expect { options.set_params({}) }.to change { options.params }.from(previous_params).to({})
    end
  end
end


```


Test case `'should be able to make params empty' # (1)` fails.

```ruby
Failures:

  1) Ably::Models::ChannelOptions#set_params should be able to make params empty
     Failure/Error: orig_handle_failure(*args, &block)
       expected `options.params` to have changed from {:example_attribute=>1} to {}, but did not change
     # ./spec/support/event_machine_helper.rb:156:in `handle_failure'
     # ./spec/lib/unit/models/channel_options_spec.rb:49:in `block (3 levels) in <top (required)>'
     # ./spec/support/rest_testapp_before_retry.rb:6:in `block (2 levels) in <top (required)>'
```


The condition mentioned here https://github.com/ably/ably-ruby/pull/336#discussion_r759455996 should be removed.

Reply:
@lukaszsliwa thanks! Really appreciate the spec 😄